core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorViewDefinition.java

```diff
@@ -56,6 +56,9 @@ public ConnectorViewDefinition(
         if (runAsInvoker && owner.isPresent()) {
             throw new IllegalArgumentException("owner cannot be present with runAsInvoker");
         }
+        if (!runAsInvoker && owner.isEmpty()) {
+            throw new IllegalArgumentException("owner must be present when runAsInvoker is false");
```

Comment:
in the spirit of the message above: "owner cannot be absent without runAsInvoker"

Reply:
BTW it's weird we're doing these checks here and not in the Analyzer (https://trinodb.slack.com/archives/CP1MUNEUX/p1617800365203200), but that's pre-existing.

Which is weird, combined with the fact `owner` was _optional_ also before `runAsInvoker` was added. 
I think today the `owner.isEmpty()` is used drive the "SECURITY INVOKER" logic in StatementAnalyzer, instead of the explicit check, but i may be missing something.

cc @martint @kokosing  @electrum

Reply:
I find it weird for the message to be defined a negative.  I believe the problem is `owner must be defined with a view with SECURITY DEFINER mode`

Reply:
This check will break deserialization of older objects that were created without an owner. Historically, the `owner` field was used to determine the security mode for views. Commit history:

* Add owner to view definition: b438df6ecb7e977aefc4017d3c86d0d136eec277
* Check view access using owner: cab3481755075d552c5a66767bf7b3d38c2c54d4
* Add explicit security mode: c408a78244ac0f9c1360b872b9569dc29c71b62f
* Expose view definition to connectors: 6efec400caf267f95c39515eacb25b69c5e735bc

This check is semantically valid from the point of view of objects returned to the engine. However, it's legitimate for a connector to deserialize a view definition and then fill in the owner field. The Hive connector does this using the table owner from the metastore.

I think the problem is that the `runAsInvoker` flag was introduced with `ViewDefinition` was an opaque serialization object in the engine and `ConnectorViewDefinition` was returned separately from the connector and included an owner field. This allowed the Hive connector to specify the view owner as the table owner from the metastore, which it did unconditionally, since the security mode was opaque to the connector.

If we had exposed view definitions to connectors at the same time as we introduced the security mode, then we could have simply used the presence of the owner field as the flag for the security mode.

So now we're left with a situation where `owner=Empty` and `runAsInvoker=false`, which can happen for legacy stored views. I think the answer is that the connector needs to sanitize this before returning it to the engine, either by adding an owner, or by changing the security mode. But we can't enforce this in the constructor, as this would break JSON deserialization.

One potential solution, to make the API more clear, is to document this and add factory methods to create both kinds of views, with the owner being required (not `Optional`) for definer security mode.

Reply:
> So now we're left with a situation where owner=Empty and runAsInvoker=false, which can happen for legacy stored views. I think the answer is that the connector needs to sanitize this before returning it to the engine, either by adding an owner, or by changing the security mode. But we can't enforce this in the constructor, as this would break JSON deserialization.

Sounds serious. Do we have a test for that?